src/uct/ib/rc/accel/rc_mlx5_devx.c

```diff
@@ -70,53 +114,97 @@ uct_rc_mlx5_devx_init_rx_tm(uct_rc_mlx5_iface_common_t *iface,
     UCT_IB_MLX5DV_SET(xrqc, xrqc, dc,       dc);
     UCT_IB_MLX5DV_SET(xrqc, xrqc, cqn,      dvcq.cqn);
 
-    wq = UCT_IB_MLX5DV_ADDR_OF(xrqc, xrqc, wq);
+    status = uct_rc_mlx5_devx_init_rx_common(iface, md, config, &dvpd,
+                                             UCT_IB_MLX5DV_ADDR_OF(xrqc, xrqc, wq));
+    if (status != UCS_OK) {
+        return UCS_OK;
+    }
 
-    UCT_IB_MLX5DV_SET(wq, wq, wq_type, iface->rx.srq.topo);
-    UCT_IB_MLX5DV_SET(wq, wq, log_wq_sz,     ucs_ilog2(max));
-    UCT_IB_MLX5DV_SET(wq, wq, log_wq_stride, ucs_ilog2(stride));
-    UCT_IB_MLX5DV_SET(wq, wq, pd,            dvpd.pdn);
-    UCT_IB_MLX5DV_SET(wq, wq, dbr_umem_id,   iface->rx.srq.devx.dbrec->mem_id);
-    UCT_IB_MLX5DV_SET64(wq, wq, dbr_addr,    iface->rx.srq.devx.dbrec->offset);
-    UCT_IB_MLX5DV_SET(wq, wq, wq_umem_id,    iface->rx.srq.devx.mem->umem_id);
+    iface->rx.srq.devx.obj = mlx5dv_devx_obj_create(dev->ibv_context,
+                                                    in, sizeof(in),
+                                                    out, sizeof(out));
+    if (iface->rx.srq.devx.obj == NULL) {
+        ucs_error("mlx5dv_devx_obj_create(XRQ) failed, syndrome %x: %m",
+                  UCT_IB_MLX5DV_GET(create_xrq_out, out, syndrome));
+        status = UCS_ERR_IO_ERROR;
+        goto err_cleanup_srq;
+    }
 
-    if (UCT_RC_MLX5_MP_ENABLED(iface)) {
-        /* Normalize to device's interface values (range of (-6) - 7) */
-        log_num_of_strides           = ucs_ilog2(iface->tm.mp.num_strides) - 9;
+    iface->rx.srq.srq_num = UCT_IB_MLX5DV_GET(create_xrq_out, out, xrqn);
+    uct_rc_mlx5_iface_tm_set_cmd_qp_len(iface);
 
-        UCT_IB_MLX5DV_SET(wq, wq, log_wqe_num_of_strides, log_num_of_strides & 0xF);
-        UCT_IB_MLX5DV_SET(wq, wq, log_wqe_stride_size,
-                          (ucs_ilog2(iface->super.super.config.seg_size) - 6));
+    return UCS_OK;
+
+err_cleanup_srq:
+    uct_rc_mlx5_devx_cleanup_srq(&iface->rx.srq);
+    return status;
+}
+
+int uct_rc_mlx5_is_devx_srq_supported(const uct_ib_mlx5_md_t *md, int dc)
+{
+    int devx_srq_requested = (dc ?
+                              (md->flags & UCT_IB_MLX5_MD_FLAG_DEVX_DC_SRQ) :
+                              (md->flags & UCT_IB_MLX5_MD_FLAG_DEVX_RC_SRQ));
+    return devx_srq_requested && (md->flags & UCT_IB_MLX5_MD_FLAG_RMP);
+}
+
+ucs_status_t uct_rc_mlx5_devx_init_rx(uct_rc_mlx5_iface_common_t *iface,
+                                      const uct_rc_iface_common_config_t *config)
+{
+    uct_ib_mlx5_md_t *md  = ucs_derived_of(uct_ib_iface_md(&iface->super.super),
+                                           uct_ib_mlx5_md_t);
+    uct_ib_device_t *dev  = &md->super.dev;
+    struct mlx5dv_pd dvpd = {};
+    struct mlx5dv_obj dv  = {};
+    char in[UCT_IB_MLX5DV_ST_SZ_BYTES(create_rmp_in)]   = {};
```

Comment:
where do we specify CQ for RMP?

Reply:
there is no CQ number field in RMP context

Reply:
right, XRQ needs cq for offloads